orttraining/orttraining/python/training/torchdynamo/ort_backend.py

```diff
@@ -0,0 +1,474 @@
+    devices: List[ORTC.OrtDevice] = []
+    first_device = values[0].device
+    for value in values:
+        assert value.device == first_device, "All values must be on the same device"
+        device_id = value.device.index or 0
+        devices.append(
+            ORTC.OrtDevice(
+                _get_ort_device_type(value.device.type, device_id), ORTC.OrtDevice.default_memory(), device_id
+            )
+        )
+    return devices
+
+
+def _run_onnx_session_with_ortvaluevector(
+    sess: onnxruntime.InferenceSession,
+    input_names: List[str],
+    inputs: Tuple[torch.Tensor],
+    input_devices: List[ORTC.OrtDevice],
+    output_names: List[str],
+    outputs: Tuple[torch.Tensor],
+    output_devices: List[ORTC.OrtDevice],
+):
+    _nvtx_range_push("contiguous")
+    inputs = tuple(a.contiguous() for a in inputs)
+    _nvtx_range_pop()
+
+    _nvtx_range_push("push_back_batch")
+    ort_inputs = ORTC.OrtValueVector()
+    ort_inputs.reserve(len(inputs))
+    ort_outputs = ORTC.OrtValueVector()
+    dtypes = []
+    shapes = []
+    data_ptrs = []
+
+    for value in inputs:
+        dtypes.append(_NP_DTYPE[value.dtype])
+        shapes.append(value.size())
+        data_ptrs.append(value.data_ptr())
+    ort_inputs.push_back_batch(inputs, data_ptrs, dtypes, shapes, input_devices)
+    _nvtx_range_pop()
+
+    _nvtx_range_push("run_with_ortvaluevector")
+    run_options = onnxruntime.RunOptions()
+    run_options.synchronize_execution_providers = False
+    sess.run_with_ortvaluevector(run_options, input_names, ort_inputs, output_names, ort_outputs, output_devices)
+    _nvtx_range_pop()
+
+    _nvtx_range_push("after run_with_ortvaluevector")
+    pth_outputs = onnxruntime.training.ortmodule._utils._ortvalues_to_torch_tensor(ort_outputs)
+    _nvtx_range_pop()
+    return pth_outputs
+
+
+def _assert_allclose_with_detailed_error_message(
+    x: torch.Tensor, y: torch.Tensor, rtol: float = 1e-03, atol: float = 1e-04
+):
+    diff = x - y
+    real_atol = torch.max(torch.abs(diff))
+    max_value = torch.max(torch.abs(x), torch.abs(y))
+    max_value[max_value == 0.0] = 1.0
+    real_rtol = torch.max(diff / max_value)
+    allclose = True if real_atol <= atol or real_rtol <= rtol else False
+    if not allclose:
+        raise RuntimeError(
+            "ONNX output doesn't match baseline output with "
+            + f"actual rtol={real_rtol} and actual atol={real_atol} "
+            + f"but expected rtol={rtol} and expected atol={atol}."
+        )
+
+
+class OrtBackend:
+    def __init__(self):
+        self.supported_ops = OrtOperatorSupport()
+        # TODO: this is a naive implementation of cache without proper guard
+        self.partitioner_cache: Dict[torch.fx.GraphModule, torch.fx.GraphModule] = {}
+        self.prim_outputs = {}
+        # TODO: this is a naive implementation of cache without proper guard, this will only work for identical inputs
+        self.onnx_sessions = {}
```

Comment:
Are multiple onnx sessions needed because the original torch model can be broken into multiple fx graphs, for ifs, loops, unsupported ops, etc?

Reply:
Yes, one unsupported op incurs one graph break.